src/client.js

```diff
@@ -320,6 +323,17 @@ export class Saturn {
     }
 
     if (lastError) {
+      const originUrl = opts.customerFallbackURL ?? this.opts.customerFallbackURL
```

Comment:
We can't assume the format of the customer fallback url. The cid could be in the subdomain, or `/ipfs/<cid>`, or `/<cid>`, or anything else. When we fallback we have to use the full URL as is, we can't assume the URL structure on line 412.

Reply:
We should also think about how this affects logging, if we fallback then what gets submitted in the log?

Reply:
We should also strip the `jwt` query param if we fallback

Reply:
> We can't assume the format of the customer fallback url.
I think initially we can just put the responsibility of that URL on the customer. They should just provide us the path where we can just put a `/<cid>` at the end and query that. 

> if we fallback then what gets submitted in the log?
We should come up with a log format for origin server requests. Are logs are pretty centered around node ids. We assign unique nodeids to client origins or we can add some info in a column which indicates that the request came from an origin server fallback. Otherwise, all the other fields can would still be useful (network error, ttfb, etc.)

Reply:
I feel like we should introduce the "fallback" option then since It seems like we are gonna have specific handling for customer origin fallbacks outside of the file type.

Reply:
> I think initially we can just put the responsibility of that URL on the customer. They should just provide us the path where we can just put a /<cid> at the end and query that.

The customer doesn't have control over configuring the fallback, or configuring the js-client in general besides the clientKey. Maybe they will in the future, but for now all they can configure is the clientKey.

Their JS/HTML will make requests, then the browser-client inspects the URL, finds the cid, and passes it to js-client. So if their requests have subdomain cids, its on us to work around that imo. Their site could also request cids from multiple origins with different URL structures, so expecting them to configure every possible fallback is a bit onerous.

> We should come up with a log format for origin server requests. Are logs are pretty centered around node ids. We assign unique nodeids to client origins or we can add some info in a column which indicates that the request came from an origin server fallback. Otherwise, all the other fields can would still be useful (network error, ttfb, etc.)

There are a few db column type requirements that we'll need to workaround, like the nodeId has to be a UUID4. 

> I feel like we should introduce the "fallback" option then since It seems like we are gonna have specific handling for customer origin fallbacks outside of the file type.

You mean like an object with several fields?

Reply:
> The customer doesn't have control over configuring the fallback, or configuring the js-client in general besides the clientKey. Maybe they will in the future, but for now all they can configure is the clientKey.

I see, so we have to handle those scenarios then. Currently, the browser client intercepts the original request and it parses the cid from the request so we can turn it into a saturn request. So would it just make sense for the service worker to pass the original request URL to the js-client and the js client just makes the fetch request for that URL without modification? 

Also you bring up another problem of the clients potentially making requests with different URL structures.  Then do we have to modify the client to parse the CID whenever a new customer has a unique URL format that we don't support? Doesn't seem like a huge issue now as we can manually do that with the first few customers. 


> Ideally the fallback log can be integrated with the bandwidth_logs table. There are a few db column type requirements that we'll need to workaround, like the nodeId has to be a UUID4.

Yes agreed. A dirty solution could be assigning a UUID4 node id to each customer we onboard and we can map that to our customer DB. That way we can easily aggregate and filter customer requests. Otherwise we can add/ use another column in the DB to mark that this request was from a fallback origin. 

In addition, trace ids can help us correlate which nodes failed before going the origin. 

> You mean like an object with several fields?

Mean that the `opts` object in the js-client for the fetch related functions should have a boolean `fallback` field to mark that the current request is an origin fallback request. That way we can ignore jwt, have separate parsing for the url, etc.

Reply:
> So would it just make sense for the service worker to pass the original request URL to the js-client and the js client just makes the fetch request for that URL without modification?

yep.

> Also you bring up another problem of the clients potentially making requests with different URL structures. Then do we have to modify the client to parse the CID whenever a new customer has a unique URL format that we don't support? Doesn't seem like a huge issue now as we can manually do that with the first few customers.

Yes, for now I've only seen subdomain cids and path cids, both of which should be detected by the browser-client https://github.com/filecoin-saturn/browser-client/blob/4281ce8f750bb06891d50c036052d71c7af99246/src/sw/controller.js#L70

> Yes agreed. A dirty solution could be assigning a UUID4 node id to each customer we onboard and we can map that to our customer DB. That way we can easily aggregate and filter customer requests. Otherwise we can add/ use another column in the DB to mark that this request was from a fallback origin.

We can probably just use the client key since its a UUID4 already.

> Mean that the opts object in the js-client for the fetch related functions should have a boolean fallback field to mark that the current request is an origin fallback request. That way we can ignore jwt, have separate parsing for the url, etc.

ah, yeah that should work.